dali/operators/random/normal_distribution_op.cu

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/random/normal_distribution_op.cuh"
+#include <vector>
+#include <utility>
+#include "dali/core/convert.h"
+
+namespace dali {
+
+namespace detail {
+
+template <typename Out, typename RandType>
+__global__ void NormalDistKernel(NormalDistributionGpu::BlockDesc *block_descs,
+                                 RandomizerGPU randomizer) {
+  auto desc = block_descs[blockIdx.x];
+  auto out = static_cast<Out*>(desc.sample);
+  auto tid = blockIdx.x * blockDim.x + threadIdx.x;
+  for (int x = desc.start + threadIdx.x; x < desc.end; x += blockDim.x) {
+    auto norm = randomizer.normal<RandType>(tid);
+    out[x] = ConvertSat<Out>(norm * desc.std + desc.mean);
+  }
+}
+
+template <typename Out, typename RandType>
+__global__ void NormalDistSingleValue(NormalDistributionGpu::BlockDesc *descs,
+                                      int size, RandomizerGPU randomizer) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid >= size) return;
+  auto desc = descs[tid];
+  auto norm = randomizer.normal<RandType>(tid);
+  *(static_cast<Out*>(desc.sample)) = ConvertSat<Out>(norm * desc.std + desc.mean);
+}
+
+std::pair<std::vector<int>, int> DistributeBlocksPerSample(
+    const TensorListShape<> &shape, int block_size, int max_blocks) {
+  std::vector<int> sizes(shape.size());
+  int sum = 0;
+  for (int i = 0; i < shape.size(); ++i) {
+    sizes[i] = div_ceil(volume(shape.tensor_shape_span(i)), block_size);
+    sum += sizes[i];
+  }
+  if (sum <= max_blocks) {
+    return {sizes, sum};
+  }
+  // If numbers of blocks exceeded max_blocks, we need to scale them down
+  int to_distribute = max_blocks - shape.size();
+  for (int i = 0; i < shape.size(); ++i) {
+    auto before = sizes[i];
+    auto scaled = before * to_distribute / sum;
+    to_distribute -= scaled;
```

Comment:
```suggestion

```
As I understand `to_distribute / sum` tells how much % you need to reduce the number of blocks in each sample. You shouldn't touch this across the loop.